packages/dds/tree/src/test/cursorTestSuite.ts

```diff
@@ -118,180 +119,330 @@ export const cursorTestCases: [string, JsonableTree][] = [
  * @param factory - Creates the cursor to be tested with or without provided data.
  * @param dataFromCursor - Gets a JsonableTree from the provided cursor.
  */
-export function testJsonableTreeCursor<T>(
+export function testJsonableTreeCursor<TData, TCursor extends ITreeCursor>(
     cursorName: string,
-    cursorFactory: (data: T) => ITreeCursor,
-    dataFromCursor: (cursor: ITreeCursor) => T,
-    testRooted = true,
+    cursorFactory: (data: TData) => TCursor,
+    dataFromCursor: (cursor: ITreeCursor) => TData,
+    extraRoot?: true,
 ): void {
-    function dataFromJsonableTree(data: JsonableTree): T {
+    function dataFromJsonableTree(data: JsonableTree): TData {
         // Use text cursor to provide input data
         return dataFromCursor(singleTextCursor(data));
     }
 
-    function factory(data: JsonableTree): ITreeCursor {
-        return cursorFactory(dataFromJsonableTree(data));
-    }
-    describe(`${cursorName} cursor implementation`, () => {
+    testTreeCursor<TData, TCursor>({
+        cursorName,
+        cursorFactory,
+        builder: dataFromJsonableTree,
+        dataFromCursor,
+        testData: cursorTestCases.map(([name, data]) => ({
+            name,
+            data: dataFromJsonableTree(data),
+            expected: data,
+        })),
+        extraRoot,
+    });
+}
+
+interface SpecialCaseBuilder<TData> {
+    withLocalKeys?(keys: LocalFieldKey[]): TData;
+    withKeys?(keys: FieldKey[]): TData;
+}
+
+/**
+ * Test suite for cursor implementations.
+ * Much of this functionality depends on dataFromJsonableTree, but basic testing can function without it.
+ */
+export function testTreeCursor<TData, TCursor extends ITreeCursor>(config: {
+    cursorName: string;
+    builder: SpecialCaseBuilder<TData> | ((data: JsonableTree) => TData);
```

Comment:
Could you add a comment about the different kinds of builders and the situations they're used in?

Reply:
Should be more clear now.